### PR TITLE
Simplify store

### PR DESCRIPTION
**Pull request by ethanwharris**
## What does this PR do?

First of several pieces to simplify the store:
- use client
- remove source code saving
- auto-detect ids + keys
- add listing
- more reasonable chunk size in download (faster downloads)

In follow-ups will add more docs, verification of user input, and other conveniences.

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

cc @carmocca @borda

### CODE REVIEW

File: .github/workflows/ci-tests-store.yml [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+  store-cpu:
+    runs-on: ${{ matrix.os }}
+    if: github.event.pull_request.draft == false
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+          - {os: "macOS-11", pkg-name: "lightning", python-version: "3.10", pytorch-version: "2.0"}
+          - {os: "ubuntu-20.04", pkg-name: "lightning", python-version: "3.10", pytorch-version: "2.0"}
+          - {os: "windows-2022", pkg-name: "lightning", python-version: "3.10", pytorch-version: "2.0"}
```
Comment (Borda) [inline .github/workflows/ci-tests-store.yml:39]:
```suggestion
        os: ["macOS-11", "ubuntu-20.04", "windows-2022"]
        pkg-name: ["lightning"]
        python-version: ["3.10"]
        pytorch-version: ["2.0"]
```

File: .github/workflows/ci-tests-store.yml [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+          # - {os: "macOS-11", pkg-name: "lightning", python-version: "3.8", pytorch-version: "2.0", requires: "oldest"}
+          # - {os: "ubuntu-20.04", pkg-name: "lightning", python-version: "3.8", pytorch-version: "2.0", requires: "oldest"}
+          # - {os: "windows-2022", pkg-name: "lightning", python-version: "3.8", pytorch-version: "2.0", requires: "oldest"}
+    timeout-minutes: 25  # because of building grpcio on Mac
+    env:
+      PACKAGE_NAME: ${{ matrix.pkg-name }}
+      FREEZE_REQUIREMENTS: ${{ ! (github.ref == 'refs/heads/master' || startsWith(github.ref, 'refs/heads/release/')) }}
+      # PYPI_CACHE_DIR: "_pip-wheels"
+      TORCH_URL_STABLE: "https://download.pytorch.org/whl/cpu/torch_stable.html"
+      TORCH_URL_TEST: "https://download.pytorch.org/whl/test/cpu/torch_test.html"
```
Comment (Borda) [inline .github/workflows/ci-tests-store.yml:50]: do you need this?
```suggestion
```

File: .github/workflows/ci-tests-store.yml [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        python-version: ${{ matrix.python-version }}
+
+    - name: basic setup
+      run: pip install -q -r .actions/requirements.txt
+
+    - name: Set min. dependencies
+      if: ${{ matrix.requires == 'oldest' }}
+      run: |
+        python .actions/assistant.py replace_oldest_ver
+
```
Comment (Borda) [inline .github/workflows/ci-tests-store.yml:66]:
```suggestion
```

File: .github/workflows/ci-tests-store.yml [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+      fail-fast: false
+      matrix:
+        include:
+          - {os: "macOS-11", pkg-name: "lightning", python-version: "3.10", pytorch-version: "2.0"}
+          - {os: "ubuntu-20.04", pkg-name: "lightning", python-version: "3.10", pytorch-version: "2.0"}
+          - {os: "windows-2022", pkg-name: "lightning", python-version: "3.10", pytorch-version: "2.0"}
+          # "oldest" versions tests, only on minimum Python
+          # - {os: "macOS-11", pkg-name: "lightning", python-version: "3.8", pytorch-version: "2.0", requires: "oldest"}
+          # - {os: "ubuntu-20.04", pkg-name: "lightning", python-version: "3.8", pytorch-version: "2.0", requires: "oldest"}
+          # - {os: "windows-2022", pkg-name: "lightning", python-version: "3.8", pytorch-version: "2.0", requires: "oldest"}
```
Comment (justusschock) [inline .github/workflows/ci-tests-store.yml:43]: nit
```suggestion
```

File: .github/workflows/ci-tests-store.yml [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    #   uses: actions/cache/restore@v3
+    #   with:
+    #     path: ${{ env.PYPI_CACHE_DIR }}
+    #     key: pypi_wheels
+    # - run: |
+    #     mkdir -p $PYPI_CACHE_DIR
+    #     ls -lh $PYPI_CACHE_DIR
+
+    # removing torch stable line:
+    # pip install -e ".[${extra}test]" "pytest-timeout" -U -f ${TORCH_URL} ${TORCH_PREINSTALL} -f ${PYPI_CACHE_DIR} --prefer-binary
```
Comment (justusschock) [inline .github/workflows/ci-tests-store.yml:77]: nit :)
```suggestion
```

File: .github/workflows/ci-tests-store.yml [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+      fail-fast: false
+      matrix:
+        os: ["macOS-11", "ubuntu-20.04", "windows-2022"]
+        pkg-name: ["lightning"]
+        python-version: ["3.10"]
+        pytorch-version: ["2.0"]
+          # "oldest" versions tests, only on minimum Python
+          # - {os: "macOS-11", pkg-name: "lightning", python-version: "3.8", pytorch-version: "2.0", requires: "oldest"}
+          # - {os: "ubuntu-20.04", pkg-name: "lightning", python-version: "3.8", pytorch-version: "2.0", requires: "oldest"}
+          # - {os: "windows-2022", pkg-name: "lightning", python-version: "3.8", pytorch-version: "2.0", requires: "oldest"}
```
Comment (Borda) [inline .github/workflows/ci-tests-store.yml:43]:
```suggestion
```

File: src/lightning/store/store.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        path:
+            The path to the checkpoint to be uploaded.
+        version:
+            The version of the model to be uploaded. If not provided, default will be latest (not overridden).
+        progress_bar:
+            A progress bar to show the uploading status. Disable this if not needed, by setting to `False`.
+    """
+    client = _Client()
+    user = client.auth_service_get_user()
+    # TODO: Allow passing this
```
Comment (justusschock) [inline src/lightning/store/store.py:43]: are you planning to resolve this in this PR or a follow-up? This seems somewhat essential.

**Review by justusschock — APPROVED** [review]

File: src/lightning/store/store.py (hunk 7 above)
Comment (ethanwharris) [reply src/lightning/store/store.py:43]: Yeah, will get to it in a follow-up

**Review by Borda — APPROVED** [review]

File: src/lightning/store/README.md [hunk 8]
```diff
@@ -13,87 +12,30 @@
 
 ```python
 import lightning as L
-from sample.model import LitAutoEncoder, Encoder, Decoder
 
-# Initialize your model here
-autoencoder = LitAutoEncoder(Encoder(), Decoder())
+# Upload a checkpoint:
+L.store.upload_model("mnist_model", "mnist_model.ckpt")
```
Comment (lantiga) [inline src/lightning/store/README.md:17]: Several frameworks (e.g. transformers) chunk up the checkpoint into multiple files.
Also, several models come with a model checkpoint and a tokenizer checkpoint.

We need to enable uploading multiple files under a single checkpoint name.

File: src/lightning/store/store.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+    # TODO: Post model parts if the file size is over threshold
+    body = V1UploadModelRequest(
+        name=f"{user.username}/{name}",
+        version=version,
+        project_id=project_id,
+    )
+    model = client.models_store_upload_model(body)
+
+    _upload_file_to_url(model.upload_url, path, progress_bar=progress_bar)
```
Comment (lantiga) [inline src/lightning/store/store.py:54]: See comment above about mapping multiple files to a single model key

**Review by lantiga — APPROVED** [review]
Added comment about multiple files per checkpoint. This is quite essential to introduce early.

File: src/lightning/store/store.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+def list_models() -> List[V1Model]:
+    """List your models in the lightning cloud.
+
+    Returns:
+        A list of model objects.
+    """
+    client = _Client()
+    # TODO: Allow passing this
+    project_id = _get_project(client).project_id
```
Comment (lantiga) [inline src/lightning/store/store.py:88]: I think we need to review the scope where checkpoints live:
if that is a project, then does the prefix used to get the model need to reflect that?
Or will I use the name of the user as a prefix? And will that allow me to get whatever models in a project that user has access to?

Also, should we use ts in lieu of project?

Comment (ethanwharris) [reply src/lightning/store/store.py:88]: Right now the prefix isn't needed, but you would need to provide the correct project ID for listing in order to see that model. If the model is private (not currently supported), then you would need to have access to that project to be able to download / list

**Pull request merged**